File: utilize-buffers/utilize.js
#!/usr/bin/env node

const path = require("path");
const fs = require("fs");
const stream = require("stream");

// Grab user input from terminal.
const infile = process.argv[2];
const outfile = "re-" + infile;

if (path.extname(infile) !== ".txt") {
    const err = new Error("input must be a plain text file -> [file].txt");
    err.evidence = infile;
    throw err;
}

const readStream = fs.createReadStream(infile, "utf-8");
readStream.on("error", (err) => console.error(err));
readStream.on("ready", function () {
    // replace(string) -> string
    // replaces 'utilize' and its variants with 'use' and its variants.
    const replace = (function () {
        const dictionary = {
            utilize: "use",
            utilise: "use",
            Utilize: "Use",
            Utilise: "Use",
            utilizes: "uses",
            utilises: "uses",
            utilized: "used",
            utilised: "used",
            utilizing: "using",
            utilising: "using",
            Utilizing: "Using",
            Utilising: "Using",
        };
        return function (text) {
            if (typeof text !== "string") {
                throw new TypeError("argument must be of type string");
            }
            return text.replace(
                /[uU]tili([zs]e|[zs]ed|[zs]ing)/g,
                function (match) {
                    return dictionary[match];
                }
            );
        };
    })();
    // Simplified constructor approach. Alternative: ES6-style constructor
    // First argument to callback function is ignored for buffers.
    // Second argument to callback function will be forwarded to the transform.push() method.
    const transform_text = new stream.Transform({
        decodeStrings: false,
        transform: function (chunk, encoding, callback) {
            callback(null, replace(chunk));
        },
    });

    const writeStream = fs.createWriteStream(outfile);
    writeStream.on("error", (err) => console.error(err));
    writeStream.on("ready", function () {
        readStream.pipe(transform_text).pipe(writeStream);
    });
    writeStream.on("finish", function() {
        console.log("Text transformation complete. Check file %s", outfile);
    });
});

File: utilize-buffers/utilize.py
#!/usr/bin/env python3
import sys, os, io, re, resource

def create():
    # convert to byte strings for fast file processing.
    dictionary = {
        b"utilize": b"use",
        b"utilise": b"use",
        b"Utilize": b"Use",
        b"Utilise": b"Use",
        b"utilizes": b"uses",
        b"utilises": b"uses",
        b"utilized": b"used",
        b"utilised": b"used",
        b"utilizing": b"using",
        b"utilising": b"using",
        b"Utilizing": b"Using",
        b"Utilising": b"Using"
    }
    # helper functions in regular expressions take match objects, not strings or bytes
    matcher = lambda mo: dictionary[mo.group(0)]

    pattern = re.compile(rb"[uU]tili(?:[zs]ed|[zs]e|[zs]ing)")

    def replacer(text):
        if type(text) != bytes:
            raise TypeError("argument must be of type bytes")
        else:
            return pattern.sub(matcher, text)

    return replacer

replace = create()
"""Scan byte string, replacing 'utilize' and its variants with 'use' and its variants."""

# 1. collect and validate user input
if __name__ == "__main__":

    file = input("Input <plain>.txt file: ")
    root, ext = os.path.splitext(file)

    if not os.path.exists(file):
        print(f"file {root} does not exist")
        sys.exit(1)

    if not ext == ".txt":
        print(f"file {root} must have '.txt' extension")
        sys.exit(1)

# 2. open files / buffer and transform data / close files
    with io.open(file, mode="rb") as text:
        with io.open("re-"+file, mode="wb") as newtext:
            n = newtext.write(replace(text.read()))
            print(f"{n} bytes written")
